### PR TITLE
[portinglayer] laundrywasher implementation

### PR DESCRIPTION
**Pull request by xshuqun**
* add laundrywasher example in portinglayer

### CODE REVIEW

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+void GenericOperationalStateDelegateImpl::HandlePauseStateCallback(GenericOperationalError & err)
+{
+    // placeholder implementation
+    auto error = GetInstance()->SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kPaused));
+    if (error == CHIP_NO_ERROR)
+    {
+        err.Set(to_underlying(ErrorStateEnum::kNoError));
+    }
+    else
```
Comment (step0035) [inline component/common/application/matter/driver/opstate_driver.cpp:54]: testing

File: component/common/application/matter/example/laundrywasher/matter_drivers.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        }
+        break;
+    default:
+        break;
+    }
+exit:
+    return;
+}
+
+void matter_driver_downlink_update_handler(AppEvent * event)
```
Comment (SirRu24) [inline component/common/application/matter/example/laundrywasher/matter_drivers.cpp:114]: Update pointer format to int *ptr

File: component/common/application/matter/example/laundrywasher/matter_drivers.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    downlink_event.mHandler = matter_driver_downlink_update_handler;
+    PostDownlinkEvent(&downlink_event);
+}
+
+void matter_driver_set_mode_callback(uint32_t id)
+{
+    AppEvent downlink_event;
+    downlink_event.Type     = AppEvent::kEventType_Downlink_LW_Mode;
+    downlink_event.value._u8 = (uint8_t) id; // 0: Stop; 1:Running ,2:Paused; 3: Error
+    downlink_event.mHandler = matter_driver_downlink_update_handler;
```
Comment (SirRu24) [inline component/common/application/matter/example/laundrywasher/matter_drivers.cpp:67]: Equal signs are not aligned for matter_driver_set_opstate_callback, matter_driver_set_spinspeed_callback, matter_driver_set_rinses_callback, and matter_driver_set_mode_callback

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    gOperationalStateInstance->SetOperationalState(to_underlying(OperationalState::OperationalStateEnum::kStopped));
+
+    gOperationalStateInstance->Init();
+}
+
+// Init RVC Operational State cluster
+
+static OperationalState::Instance * gRvcOperationalStateInstance  = nullptr;
+static RvcOperationalStateDelegate * gRvcOperationalStateDelegate = nullptr;
```
Comment (SirRu24) [inline component/common/application/matter/driver/opstate_driver.cpp:143]: Line 104, 105, 121, 142, 143: Update pointer declaration to int *ptr format

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        err.Set(to_underlying(ErrorStateEnum::kNoError));
+    }
+    else
+    {
+        err.Set(to_underlying(ErrorStateEnum::kUnableToCompleteOperation));
+    }
+}
+
+void GenericOperationalStateDelegateImpl::HandleStopStateCallback(GenericOperationalError & err)
```
Comment (SirRu24) [inline component/common/application/matter/driver/opstate_driver.cpp:88]: Line 26, 36, 46, 60, 74, 88: update reference declaration style to GenericOperationalError &err format

File: component/common/application/matter/driver/opstate_driver.h [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+void getOperationalStateSet(u8 state);
+
+namespace chip {
+namespace app {
+namespace Clusters {
+
+namespace OperationalState {
+
+    Instance* GetOperationalStateInstance();
```
Comment (SirRu24) [inline component/common/application/matter/driver/opstate_driver.h:34]: update pointer style to int *ptr format

File: component/common/application/matter/driver/opstate_driver.h [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+     * Handle Command Callback in application: Start
+     * @param[out] get operational error after callback.
+     */
+    void HandleStartStateCallback(GenericOperationalError & err) override;
+
+    /**
+     * Handle Command Callback in application: Stop
+     * @param[out] get operational error after callback.
+     */
+    void HandleStopStateCallback(GenericOperationalError & err) override;
```
Comment (SirRu24) [inline component/common/application/matter/driver/opstate_driver.h:89]: Line 71, 77, 83, 89: update reference declaration to GenericOperationalError &err format

File: component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_core.mk [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+#for time64 
+ifdef SYSTEM_TIME64_MAKE_OPTION
+CFLAGS += -DCONFIG_SYSTEM_TIME64=1
+CFLAGS += -include time64.h
+else
+CFLAGS += -DCONFIG_SYSTEM_TIME64=0
+endif
+
+# for matter blemgr adapter
+#CFLAGS += -DCONFIG_MATTER_BLEMGR_ADAPTER=1
```
Comment (SirRu24) [inline component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_core.mk:184]: Line 70, 184: clean commented out lines

File: component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_main.mk [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/coap/include
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/libcoap/include
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/http2/nghttp2-1.31.0/includes
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/lwip/lwip_v2.1.2/src/include
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/lwip/lwip_v2.1.2/src/include/lwip
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/lwip/lwip_v2.1.2/port/realtek
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/lwip/lwip_v2.1.2/port/realtek/freertos
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/ssl/mbedtls-matter/include
+INCLUDES += -I$(SDKROOTDIR)/component/common/network/ssl/mbedtls-matter/include/mbedtls
+#INCLUDES += -I$(SDKROOTDIR)/component/common/network/ssl/ssl_ram_map/rom
```
Comment (SirRu24) [inline component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_main.mk:76]: Line 14, 76: clean commented out commands

**Review by SirRu24 — COMMENTED** [review]
opstate_driver.cpp, opstate_driver.h, matter_drivers.cpp: Can change the style of pointer and reference declaration, not urgent
lib_chip_laundrywasher_core.mk, lib_chip_laundrywasher_mian.mk: Clean the commented out line, I am not sure if it was intended to be left here or forget to delete
zap file not tested yet

File: component/common/application/matter/example/laundrywasher/matter_drivers.cpp (hunk 3 above)
Comment (step0035) [reply component/common/application/matter/example/laundrywasher/matter_drivers.cpp:67]: equal sign alignment is fine, it is only enforced in connectedhomeip

File: component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_core.mk (hunk 8 above)
Comment (step0035) [reply component/common/application/matter/example/laundrywasher/lib_chip_laundrywasher_core.mk:184]: this is a special case of comment, can leave it in

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        return CHIP_ERROR_NOT_FOUND;
+    }
+    operationalPhase = mOperationalPhaseList[index];
+    return CHIP_NO_ERROR;
+}
+
+void GenericOperationalStateDelegateImpl::HandlePauseStateCallback(GenericOperationalError & err)
+{
+    // placeholder implementation
```
Comment (step0035) [inline component/common/application/matter/driver/opstate_driver.cpp:48]: can delete this placeholder implementation as this is already our own implementation, might be misleading

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    else
+    {
+        err.Set(to_underlying(ErrorStateEnum::kUnableToCompleteOperation));
+    }
+}
+
+void GenericOperationalStateDelegateImpl::HandleResumeStateCallback(GenericOperationalError & err)
+{
+    // placeholder implementation
```
Comment (step0035) [inline component/common/application/matter/driver/opstate_driver.cpp:62]: same =)

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    else
+    {
+        err.Set(to_underlying(ErrorStateEnum::kUnableToCompleteOperation));
+    }
+}
+
+void GenericOperationalStateDelegateImpl::HandleStartStateCallback(GenericOperationalError & err)
+{
+    // placeholder implementation
```
Comment (step0035) [inline component/common/application/matter/driver/opstate_driver.cpp:76]: same =))

File: component/common/application/matter/driver/opstate_driver.cpp [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    else
+    {
+        err.Set(to_underlying(ErrorStateEnum::kUnableToCompleteOperation));
+    }
+}
+
+void GenericOperationalStateDelegateImpl::HandleStopStateCallback(GenericOperationalError & err)
+{
+    // placeholder implementation
```
Comment (step0035) [inline component/common/application/matter/driver/opstate_driver.cpp:90]: =)))

**Review by step0035 — CHANGES_REQUESTED** [review]
no longer a placeholder implementation

**Review by SandyYeow — APPROVED** [review]

**Review by step0035 — APPROVED** [review]

**Review by SirRu24 — APPROVED** [review]

**Pull request merged**